Feature/init from config space (closes #36)

LGro opened the pull request:
This PR draft is the first attempt to loading config space JSONs / dictionaries into ParameterSpace instances. This should help us run BBO optimizers on benchmarks that expose config spaces. However, there isn't always a 1:1 mapping (e.g. normal vs. truncated normal prior) and I might have missed other less commonly used ConfigSpace features. To get a second pair of eyes on this early, I'd like to ask you to take a look at it and let me know if you see open issues, @sfalkner.

sfalkner commented on parameterspace/configspace_utils.py:
```diff
@@ -0,0 +1,130 @@
+                    name=param_name,
+                    bounds=(lower_bound, upper_bound),
+                    prior=ps.priors.TruncatedNormal(
+                        mean=param_dict["mu"], std=param_dict["sigma"]
```
If I am not mistaken, `ps.priors.TruncatedNormal` expects the mean and standard deviation in the normalized space. So given the bounds from above, I think it should be:
```suggestion
                        mean=0.5, std=1./8.
```

sfalkner commented on tests/test_configspace_utils.py:
```diff
@@ -0,0 +1,202 @@
+
+    param = space.get_parameter_by_name("p")["parameter"]
+    assert isinstance(param._prior, TruncatedNormalPrior)
+    assert param._prior.mean == cs_dict["hyperparameters"][0]["mu"]
```
I think a more meaningful test could be to draw samples from the parameterspace and compare the mean of those to the specified one (with some reasonable slack). That way, we would see if the TruncatedNormalPrior is initialized correctly above.

LGro commented:
I was expecting more loose ends to tie up before opening up this PR for review, but here we are. @sfalkner, I fixed the truncated normal prior and adjusted the tests to compare samples to the mean and std, which I'd invite you to check again.
I'm also inviting @lreeb for a fresh pair of eyes :relaxed:

sfalkner commented on parameterspace/configspace_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+    varnames = []
+    function_texts = []
+    for cond in conditions:
```
Not sure how quickly this becomes relevant, but there are more conditions in ConfigSpace, e.g. `GreaterThanCondition` and `LessThanCondition`. Additionally, there are also `And/OrConjunction` to combine conditions. In practice, I don't think these are super relevant as the "IN" and "NOTIN"(?) already cover most cases.

LGro replied:
Following https://automl.github.io/ConfigSpace/main/api/conditions.html there are

- ✅ EqualsCondition
- ❌ NotEqualsCondition
- ❌ LessThanCondition
- ❌ GreaterThanCondition
- ✅ InCondition
- ❌ AndConjunction
- ❌ OrConjunction

I'll integrate the missing ones since support should be trivial.

Regarding the conjunctions, I'm less sure how trivial this is because with that one can have arbitrary hierarchies of conditions, right? So I might opt to raise a `NotImplementedError` in the presence of conjunctions and test for that for now.

sfalkner replied:
The Conjunction will be non-trivial for sure. As you mentioned, they do support a hierarchy of conditions and more conjunctions, which requires some thought and is probably not used in most cases. At least for now, I don't think we really need the Conjunctions.

LGro replied:
I created #51 for additional transparency and us not to forget about it.

sfalkner commented on parameterspace/configspace_utils.py:
```diff
@@ -0,0 +1,132 @@
+            varnames.append(parent)
+            # The representation is used because it quotes strings.
+            if cond["type"] == "IN":
+                # Using the valid Python "a in [2,3,4]" is not supported, hence use or
```
I am surprised by this comment. Does that refer to the way the lambda function is constructed here, because parameterspace does support the pythonic `a in [2,3,4]` as a condition?

LGro replied:
IIRC we don't treat this as a safe statement for de-/serialization, which makes the explicitly constructed or equal statement more useful. I'd be happy to spend some time on making `in` safe for de-serialization, though.

LGro replied:
Should be resolved with #50, on which I'll rebase once merged and simplify the condition here.

lreeb reviewed (approved):
Looks really good, thanks @LGro!

LGro commented on parameterspace/configspace_utils.py:
```diff
@@ -0,0 +1,197 @@
+                ),
+                "condition": condition,
+            }
+
```
Thanks for pointing that potential out, it does seem quite reasonable. However, since we currently don't see the requirement and it's frankly not possible due to automl/ConfigSpace/issues/279 to get a matching dictionary representation at the moment other than writing it by hand, I created #52 as a reminder for us to get back to it.
